src/chainlink/ChainlinkOracle.sol

```diff
@@ -0,0 +1,215 @@
+    }
+
+    /// @inheritdoc IOracle
+    function priceOf(address base, address quote) external view returns (uint256 price) {
+        price = _valueOf(base, quote, 10 ** IERC20(base).decimals());
+    }
+
+    /// -----------------------------------------------------------------------
+    /// functions - private & internal - views
+    /// -----------------------------------------------------------------------
+
+    function _valueOf(address base, address quote, uint256 baseAmount) internal view returns (uint256) {
+        // fetch pair details
+        PairDetail memory pd = pairDetails[base][quote];
+
+        // if path is not set, base-quote is not supported
+        if (pd.path.length == 0) {
+            revert OracleUnsupportedPair(base, quote);
+        }
+
+        // decode feeds from the path
+        Feed[] memory feeds = pd.path.getFeeds();
+        uint256 feedLength = feeds.length;
+
+        uint256 price;
+        uint256 answer;
+        bool isInvalid;
+
+        // fetch price for the first feed
+        (price, isInvalid) = _getFeedAnswer(feeds[0]);
+        if (isInvalid) revert OracleUntrustedData(base, quote);
+
+        for (uint256 i = 1; i < feedLength;) {
+            (answer, isInvalid) = _getFeedAnswer(feeds[i]);
+
+            if (isInvalid) revert OracleUntrustedData(base, quote);
+
+            if (feeds[i].mul) {
+                price = price.mulWadDown(answer);
+            } else {
+                price = price.divWadDown(answer);
+            }
+            unchecked {
+                ++i;
+            }
+        }
+
+        // inverse the final price if path is inverted
+        if (pd.inverted) price = WAD.divWadDown(price);
+
+        if (price == 0) revert OracleUntrustedData(base, quote);
+
+        return _convertPriceToQuoteAmount(base, quote, baseAmount, price);
+    }
+
+    function _getFeedAnswer(Feed memory _feed) internal view returns (uint256 scaledAnswer, bool isInvalid) {
+        (
+            , /* uint80 roundId, */
+            int256 answer,
+            , /* uint256 startedAt, */
+            uint256 updatedAt,
+            /* uint80 answeredInRound */
+        ) = _feed.feed.latestRoundData();
+
+        // check for stale price
+        if (updatedAt + _feed.staleAfter < block.timestamp) {
+            isInvalid = true;
+        }
+        if (answer < 0) {
+            isInvalid = true;
+        }
+
+        // scale answert to 18 decimals
+        if (_feed.decimals <= 18) {
+            return (uint256(answer) * 10 ** (18 - _feed.decimals), isInvalid);
+        } else {
+            return (uint256(answer) / 10 ** (_feed.decimals - 18), isInvalid);
+        }
+    }
+
+    function _convertPriceToQuoteAmount(address base, address quote, uint256 baseAmount, uint256 price)
+        internal
+        view
+        returns (uint256 finalAmount)
+    {
+        uint8 baseDecimals = IERC20(base).decimals();
+        uint8 quoteDecimals = IERC20(quote).decimals();
```

Comment:
These CALLs will be a bit expensive, considering this will be an immutable contract. We should store the scalars.